data-products/src/main/scala/org/sunbird/obsrv/dataproducts/job/MasterDataProcessorIndexer.scala

```diff
@@ -0,0 +1,156 @@
+    val indexedDatasets = datasets.filter(dataset => {
+      dataset.datasetConfig.indexData.nonEmpty && dataset.datasetConfig.indexData.get
+    })
+    val metrics = BaseMetricHelper(config)
+    indexedDatasets.foreach(dataset => {
+      metrics.generate(ets = new DateTime(DateTimeZone.UTC).getMillis, datasetId = dataset.id, edata = Edata(metric = Map(metrics.getMetricName("total_dataset_count") -> 1), labels = List(MetricLabel("job", "MasterDataIndexer"), MetricLabel("datasetId", dataset.id), MetricLabel("cloud", s"${config.getString("cloudStorage.provider")}"))))
+        indexDataset(dataset, metrics, System.currentTimeMillis())
+    })
+  }
+
+  def indexDataset(dataset: Dataset, metrics: BaseMetricHelper, time: Long): Unit = {
+    try {
+      val datasources = DatasetRegistry.getDatasources(dataset.id)
+      if (datasources.isEmpty || datasources.get.size > 1) {
+        return
+      }
+      val datasource = datasources.get.head
+      val paths = getPaths(datasource)
+      val events_count = createDataFile(dataset, paths.outputFilePath)
+      val ingestionSpec = updateIngestionSpec(datasource, paths.datasourceRef, paths.ingestionPath)
+      submitIngestionTask(ingestionSpec)
+      DatasetRegistry.updateDatasourceRef(datasource, paths.datasourceRef)
+      if (!datasource.datasourceRef.equals(paths.datasourceRef)) {
+        deleteDataSource(datasource.datasourceRef)
+      }
+      val end_time = System.currentTimeMillis()
+      val success_time = end_time - time
+      metrics.generate(ets = new DateTime(DateTimeZone.UTC).getMillis, datasetId = dataset.id, edata = Edata(metric = Map(metrics.getMetricName("success_dataset_count") -> 1, metrics.getMetricName("total_time_taken") -> success_time, metrics.getMetricName("total_events_processed") -> events_count), labels = List(MetricLabel("job", "MasterDataIndexer"), MetricLabel("datasetId", dataset.id), MetricLabel("cloud", s"${config.getString("cloudStorage.provider")}"))))
+    } catch {
+      case e: Exception =>
+        metrics.generate(ets = new DateTime(DateTimeZone.UTC).getMillis, datasetId = dataset.id, edata = Edata(metric = Map(metrics.getMetricName("failure_dataset_count") -> 1), labels = List(MetricLabel("job", "MasterDataIndexer"), MetricLabel("datasetId", dataset.id), MetricLabel("cloud", s"${config.getString("cloudStorage.provider")}")), err = "Failed to index dataset.", errMsg = e.getMessage))
+        e.printStackTrace()
+    }
+  }
+
+  def getPaths(datasource: DataSource): Paths = {
+    val dt = new DateTime(DateTimeZone.UTC).withTimeAtStartOfDay()
+    val timestamp = dt.getMillis
+    val date = dayPeriodFormat.print(dt)
+    val provider = providerFormat(config.getString("cloudStorage.provider"))
+    val cloudPrefix = if (config.getString("cloudStorage.provider") == "azure") {
+      provider.sparkProviderURIFormat + s"""://${config.getString("cloudStorage.accountName")}.blob.core.windows.net/${config.getString("cloudStorage.container")}/"""
```

Comment:
What about S3 and local?

Reply:
Can we have separate methods to get the path format? Something like this - https://github.com/Sunbird-Obsrv/sunbird-analytics-core/blob/release-5.1.3/analytics-core/src/main/scala/org/ekstep/analytics/framework/util/CommonUtil.scala#L702-L725

Reply:
Ok will refer it.